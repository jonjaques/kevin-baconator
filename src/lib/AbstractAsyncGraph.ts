
export interface ISerialized<N, E> { nodes: N[], links: E[] }

export abstract class AbstractAsyncGraph<TNode, TEdge> {

  public addNode(node: TNode): Promise<TNode> {
    return this.addNodeInternal(node)
  }

  public removeNode(node: TNode): Promise<void> {
    return this.removeNodeInternal(node)
  }

  public nodes(): Promise<TNode[]> {
    return this.nodesInternal()
  }

  public adjacent(node: TNode): Promise<TNode[]> {
    return this.adjacentInternal(node)
  }

  public addEdge(u: string, v: string, weight?: number): Promise<TEdge> {
    return this.addEdgeInternal(u, v, weight)
  }

  public removeEdge(u: string, v: string): Promise<void> {
    return this.removeEdgeInternal(u, v)
  }

  public setEdgeWeight(u: string, v: string, weight: number): Promise<TEdge> {
    return this.setEdgeWeightInternal(u, v, weight)
  }

  public getEdgeWeight(u: string, v: string): Promise<number> {
    return this.getEdgeWeightInternal(u, v)
  }

  public indegree(node: TNode): Promise<number> {
    return this.indegreeInternal(node)
  }

  public outdegree(node: TNode): Promise<number> {
    return this.outdegreeInternal(node)
  }

  public depthFirstSearch(sourceNodes?: TNode[], includeSourceNodes?: boolean): Promise<TNode[]> {
    return this.depthFirstSearchInternal(sourceNodes, includeSourceNodes)
  }

  public lowestCommonAncestors(node1: TNode, node2: TNode): Promise<TNode[]> {
    return this.lowestCommonAncestorsInternal(node1, node2)
  }

  public topologicalSort(sourceNodes?: TNode[], includeSourceNodes?: boolean): Promise<TNode[]> {
    return this.topologicalSortInternal(sourceNodes, includeSourceNodes)
  }

  public shortestPath(source: TNode, destination: TNode): Promise<TNode[]> {
    return this.shortestPathInternal(source, destination)
  }

  public serialize() {
    return this.serializeInternal()
  }

  public initialize(serialized?: ISerialized<TNode, TEdge>): Promise<this> {
    return this.initializeInternal(serialized)
  }

  public deserialize(serialized: ISerialized<any, any>): Promise<this> {
    return this.deserializeInternal(serialized)
  }

  public serializeNode(node: TNode): any {
    return this.serializeNodeInternal(node)
  }

  public deserializeNode(serialized: ISerialized<TNode, TEdge>): TNode {
    return this.deserializeNodeInternal(serialized)
  }

  protected abstract addNodeInternal(node: TNode): Promise<TNode>

  protected abstract removeNodeInternal(node: TNode): Promise<void>

  protected abstract nodesInternal(): Promise<TNode[]>

  protected abstract adjacentInternal(node: TNode): Promise<TNode[]>

  protected abstract addEdgeInternal(u: string, v: string, weight?: number): Promise<TEdge>

  protected abstract removeEdgeInternal(u: string, v: string): Promise<void>

  protected abstract setEdgeWeightInternal(u: string, v: string, weight: number): Promise<TEdge>

  protected abstract getEdgeWeightInternal(u: string, v: string): Promise<number>

  protected abstract indegreeInternal(node: TNode): Promise<number>

  protected abstract outdegreeInternal(node: TNode): Promise<number>

  protected abstract depthFirstSearchInternal(sourceNodes?: TNode[], includeSourceNodes?: boolean): Promise<TNode[]>

  protected abstract lowestCommonAncestorsInternal(node1: TNode, node2: TNode): Promise<any>

  protected abstract topologicalSortInternal(sourceNodes?: TNode[], includeSourceNodes?: boolean): Promise<TNode[]>

  protected abstract shortestPathInternal(source: TNode, destination: TNode): Promise<TNode[]>

  protected abstract serializeInternal(): Promise<ISerialized<any, any>>

  protected abstract deserializeInternal(serialized: ISerialized<any, any>): Promise<this>

  protected abstract initializeInternal(serialized?: ISerialized<any, any>): Promise<this>

  protected abstract serializeNodeInternal(node: TNode): any

  protected abstract deserializeNodeInternal(serialized: any): TNode

  // Computes a string encoding of an edge,
  // for use as a key in an object.
  protected encodeEdge(u: string, v: string) {
    return u + "|" + v
  }
}
