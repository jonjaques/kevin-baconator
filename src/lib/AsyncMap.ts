export class AsyncMap<K, V> {
  private map: Map<K, V> = new Map()

  public async clear() {
    this.map.clear()
  }

  public async delete(key: K): Promise<boolean> {
    return this.map.delete(key)
  }

  public async entries() {
    return this.map.entries()
  }

  public forEach(cb: (value: V, key: K, map: Map<K, V>) => void) {
    const keys = Array.from(this.map.keys())
    return Promise.all(keys.map((k: K) => {
      return Promise.resolve(cb(this.map.get(k)!, k, this.map))
    }))
  }

  public async get(key: K): Promise<V | null> {
    return this.map.has(key) ? this.map.get(key)! : null
  }

  public async has(key: K): Promise<boolean> {
    return this.map.has(key)
  }

  public async keys(): Promise<K[]> {
    return Array.from(this.map.keys())
  }

  public async set(key: K, val: V) {
    return this.map.set(key, val)
  }

  public async size() {
    return this.map.size
  }

  public async values() {
    return this.map.values()
  }
}
