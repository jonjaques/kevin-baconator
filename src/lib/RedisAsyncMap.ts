import Redis from "redis"
import UUID from "UUID"

export class RedisAsyncMap<V> {
  private type: "list" | "key"
  private name: string
  private client = Redis.createClient()

  constructor(type: "list" | "key", name: string = UUID.v4()) {
    this.type = type
    this.name = name

  }

  public get(key: string): Promise<V | null> {
    return new Promise((resolve, reject) => {
      this.client.get(key, (err, val) => {
        if ((err === null && val === null) || val === null) {
          return resolve(null)
        }
        if (err) {
          return reject(err)
        }
        return resolve(val as unknown as V)
      })
    })
  }

  public has(key: string, ...args: string[]): Promise<boolean> {
    return new Promise((resolve, reject) => {
      this.client.exists(key, args, (err, n) => {
        if (err) {
          return reject(err)
        }
        return resolve(n > 0)
      })
    })
  }

  public async keys(): Promise<string[]> {
    return new Promise((resolve, reject) => {
      resolve([])
    })
  }

  public async set(key: string, val: V): Promise<void> {
    return new Promise((resolve, reject) => {
      resolve()
    })
  }

  public async delete(key: string): Promise<void> {
    return new Promise((resolve, reject) => {
      this.client.del(key, (err, nDel) => {

      })
    })
  }
}
