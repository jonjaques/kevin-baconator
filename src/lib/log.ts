import qs from "querystring"
import UUID from "uuid"

export function profile(name: string, data: any = {}) {
  const args = { id: UUID.v4(), ...data }
  const key = `${name}(${qs.stringify(args, "&", "=", { encodeURIComponent: (x) => x })})`
  console.time(key)
  return function end() {
    console.timeEnd(key)
  }
}
