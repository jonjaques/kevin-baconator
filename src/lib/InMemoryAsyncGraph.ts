import { AbstractAsyncGraph, ISerialized } from "./AbstractAsyncGraph"

export type Node = string

export type Edge = [string, string, number]

export interface INode { id: string }

export interface IEdge { source: string, target: string, weight?: number }

export interface IVisited { [node: string]: boolean }

export type Serialized = ISerialized<INode, IEdge>

export class InMemoryAsyncGraph extends AbstractAsyncGraph<Node, Edge> {
  private edges: Map<Node, Node[]> = new Map()
  private edgeWeights: Map<string, number> = new Map()

  public async deserialize(serialized: Serialized): Promise<this> {
    return this.deserializeInternal(serialized)
  }

  protected async initializeInternal(serialized?: Serialized) {
    if (serialized) {
      await this.deserializeInternal(serialized)
    }

    return this
  }

  protected async serializeInternal(): Promise<Serialized> {
    const nodes = await this.nodesInternal()
    const serialized = {
      links: [] as IEdge[],
      nodes: nodes.map((id) => ({ id })),
    }

    for (const source of nodes) {
      const adjacent = await this.adjacentInternal(source)
      for (const target of adjacent) {
        serialized.links.push({
          source,
          target,
          weight: await this.getEdgeWeightInternal(source, target),
        })
      }
    }

    return serialized
  }

  protected async deserializeInternal(serialized: Serialized): Promise<this> {
    for (const sNode of serialized.nodes) {
      await this.addNodeInternal(sNode.id)
    }

    for (const sEdge of serialized.links) {
      await this.addEdgeInternal(sEdge.source, sEdge.target, sEdge.weight)
    }

    return this
  }

  protected deserializeNodeInternal(node: { id: string }) {
    return node.id
  }

  protected serializeNodeInternal(node: Node) {
    return { id: node }
  }

  protected async addNodeInternal(node: Node) {
    this.edges.set(node, await this.adjacentInternal(node))
    return node
  }

  protected async removeNodeInternal(node: Node) {
    const nodes = this.edges.keys()
    for (const u of nodes) {
      for (const v of this.edges.get(u)!) {
        if (v === node) {
          await this.removeEdgeInternal(u, v)
        }
      }
    }

    this.edges.delete(node)
  }

  protected async nodesInternal() {
    const nodeSet = new Set<string>()
    for (const node of this.edges.keys()) {
      nodeSet.add(node)
      for (const edge of this.edges.get(node)!) {
        nodeSet.add(edge)
      }
    }
    return Array.from(nodeSet.keys())
  }

  protected async adjacentInternal(node: Node) {
    if (!this.edges.has(node)) {
      return []
    }
    return this.edges.get(node)!
  }

  protected async setEdgeWeightInternal(u: string, v: string, weight: number) {
    const edge = this.encodeEdge(u, v)
    this.edgeWeights.set(edge, weight)
    return [u, v, weight] as Edge
  }

  protected async getEdgeWeightInternal(u: string, v: string) {
    const edge = this.encodeEdge(u, v)
    const weight = this.edgeWeights.get(edge)
    return weight === undefined ? 1 : weight
  }

  protected async addEdgeInternal(u: string, v: string, weight?: number) {
    await this.addNodeInternal(u)
    await this.addNodeInternal(v)
    const adjacent = await this.adjacentInternal(u)
    adjacent.push(v)

    if (weight !== undefined) {
      await this.setEdgeWeightInternal(u, v, weight)
    }

    return [u, v, weight] as Edge
  }

  protected async removeEdgeInternal(u: string, v: string) {
    if (this.edges.has(u)) {
      const adjacent = await this.adjacentInternal(u)
      this.edges.set(u, adjacent.filter((f) => v !== f))
    }
  }

  protected async indegreeInternal(node: Node) {
    let degree = 0

    for (const key of this.edges.keys()) {
      for (const link of this.edges.get(key)!) {
        if (link === node) {
          degree++
        }
      }
    }

    return degree
  }

  protected async outdegreeInternal(node: Node) {
    return this.edges.has(node) ? this.edges.get(node)!.length : 0
  }

  protected async depthFirstSearchInternal(sourceNodes?: Node[], includeSourceNodes?: boolean) {
    if (!sourceNodes) {
      sourceNodes = await this.nodesInternal()
    }

    if (typeof includeSourceNodes !== "boolean") {
      includeSourceNodes = true
    }

    const visited = new Set<string>()
    const nodeList: Node[] = []

    const visit = async (node: Node) => {
      if (!visited.has(node)) {
        visited.add(node)
        const adjacent = await this.adjacentInternal(node)
        for (const adj of adjacent) {
          await visit(adj)
        }
        nodeList.push(node)
      }
    }

    if (includeSourceNodes) {
      for (const node of sourceNodes) {
        await visit(node)
      }
    } else {
      for (const node of sourceNodes) {
        visited.add(node)
      }
      for (const node of sourceNodes) {
        const adjacent = await this.adjacentInternal(node)
        for (const adj of adjacent) {
          await visit(adj)
        }
      }
    }

    return nodeList
  }

  protected async topologicalSortInternal(sourceNodes?: Node[], includeSourceNodes?: boolean) {
    const search = await this.depthFirstSearchInternal(sourceNodes, includeSourceNodes)
    return search.reverse()
  }
  protected async shortestPathInternal(source: string, destination: string): Promise<Node[]> {

    // Upper bounds for shortest path weights from source.
    const d: { [node: string]: number } = {}

    // Predecessors.
    const p: { [node: string]: string } = {}

    // Poor man's priority queue, keyed on d.
    let q: { [node: string]: boolean } = {}

    const initializeSingleSource = async () => {
      (await this.nodesInternal()).forEach((node) => {
        d[node] = Infinity
      })
      if (d[source] !== Infinity) {
        throw new Error("Source node is not in the graph")
      }
      if (d[destination] !== Infinity) {
        throw new Error("Destination node is not in the graph")
      }
      d[source] = 0
    }

    // Adds entries in q for all nodes.
    const initializePriorityQueue = async () => {
      (await this.nodesInternal()).forEach((node) => {
        q[node] = true
      })
    }

    // Returns true if q is empty.
    const priorityQueueEmpty = () => {
      return Object.keys(q).length === 0
    }

    // Linear search to extract (find and remove) min from q.
    const extractMin = () => {
      let min = Infinity
      let minNode: string | undefined
      Object.keys(q).forEach((node) => {
        if (d[node] < min) {
          min = d[node]
          minNode = node
        }
      })
      if (minNode === undefined) {
        // If we reach here, there's a disconnected subgraph, and we're done.
        q = {}
        return null
      }
      delete q[minNode]
      return minNode
    }

    const relax = async (u: string, v: string) => {
      const w = await this.getEdgeWeightInternal(u, v)
      if (d[v] > d[u] + w) {
        d[v] = d[u] + w
        p[v] = u
      }
    }

    const dijkstra = async () => {
      await initializeSingleSource()
      await initializePriorityQueue()
      while (!priorityQueueEmpty()) {
        const u = extractMin()
        if (u !== null) {
          const adj = await this.adjacentInternal(u)
          for (const v of adj) {
            await relax(u!, v)
          }
        }
      }
    }

    // Assembles the shortest path by traversing the
    // predecessor subgraph from destination to source.
    const path = async () => {
      const nodeList: any = []
      let weight = 0
      let node = destination
      while (p[node]) {
        nodeList.push(node)
        weight += await this.getEdgeWeightInternal(p[node], node)
        node = p[node]
      }
      if (node !== source) {
        throw new Error("No path found")
      }
      nodeList.push(node)
      nodeList.reverse()
      nodeList.weight = weight
      return nodeList
    }

    await dijkstra()

    return path()
  }

  protected async lowestCommonAncestorsInternal(node1: Node, node2: Node): Promise<Node[]> {
    const node1Ancestors: Node[] = []
    const lcas: Node[] = []

    const CA1Visit = async (visited: IVisited, node: string): Promise<boolean> => {
      if (!visited[node]) {
        visited[node] = true
        node1Ancestors.push(node)
        if (node === node2) {
          lcas.push(node)
          return false // found - shortcut
        }

        const adjacent = await this.adjacentInternal(node)
        const adjacentVisited = await Promise.all(adjacent.map((a) => CA1Visit(visited, a)))
        return adjacentVisited.every((a) => a === true)
      } else {
        return true
      }
    }

    const CA2Visit = async (visited: IVisited, node: Node) => {
      if (!visited[node]) {
        visited[node] = true
        if (node1Ancestors.indexOf(node) >= 0) {
          lcas.push(node)
        } else if (lcas.length === 0) {
          const adjacent = await this.adjacentInternal(node)
          await Promise.all(adjacent.map((a) => CA2Visit(visited, a)))
        }
      }
    }

    if (await CA1Visit({}, node1)) { // No shortcut worked
      await CA2Visit({}, node2)
    }

    return lcas
  }
}
