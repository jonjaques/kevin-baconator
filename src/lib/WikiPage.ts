import Wiki, { Page } from "wikijs"
import { InMemoryAsyncGraph } from "./InMemoryAsyncGraph"
import { profile } from "./log"

const wiki = Wiki()

export { Page }

export class WikiPage {

  public static fromJSON(obj: any = {}) {
    const page = new WikiPage(obj.name, obj.id)
    page.backlinks = obj.backlinks
    page.categories = obj.categories
    page.links = obj.links
    page.url = obj.url
    return page
  }

  public name: string
  public links?: string[]
  public backlinks?: string[]
  public categories?: string[]
  public id?: number
  public url?: URL
  public size?: number
  private repository?: Page

  constructor(name: string, id?: number) {
    this.name = name
    if (id) {
      this.id = id
    }
  }

  public async attachToGraph(graph: InMemoryAsyncGraph) {
    await graph.addNode(this.name)
    if (this.links) {
      await Promise.all(this.links.map((link) => graph.addEdge(this.name, link)))
    }
    if (this.backlinks) {
      await Promise.all(this.backlinks.map((link) => graph.addEdge(link, this.name)))
    }
  }

  public async fetchCategories() {
    const end = profile(`WikiPage.fetchCategories`, { name: this.name })
    if (!this.repository) {
      await this.find()
    }

    this.categories = await this.repository!.categories()
    end()
  }

  public async fetchLinks() {
    const end = profile(`WikiPage.fetchLinks`, { name: this.name })
    if (!this.repository) {
      await this.find()
    }

    this.links = await this.repository!.links()
    this.backlinks = await this.repository!.backlinks()
    end()
  }

  public async find() {
    const end = profile(`WikiPage.find`, { name: this.name })
    if (this.id) {
      this.repository = await wiki.findById(this.id as unknown as string)
    } else {
      this.repository = await wiki.find(this.name)
    }

    const r = this.repository as any
    this.id = r.raw.pageid
    this.name = r.raw.title
    this.size = r.raw.length
    this.url = this.repository.url()
    end()
  }

  public toJSON() {
    return {
      backlinks: this.backlinks,
      categories: this.categories,
      id: this.id,
      links: this.links,
      name: this.name,
      url: this.url,
    }
  }
}
