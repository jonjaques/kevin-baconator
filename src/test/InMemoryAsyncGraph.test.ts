// If using from the NPM package, this line would be
// const graph = require("graph-data-structure");
import assert from "assert"
import { InMemoryAsyncGraph as Graph, Serialized } from "../lib/InMemoryAsyncGraph"
import { comesBefore, contains, withWeight } from "./utils"

describe("Graph", () => {
  describe("Data structure", () => {
    it("Should add nodes and list them.", async () => {
      const graph = new Graph()
      await graph.addNode("a")
      await graph.addNode("b")
      expect(await graph.nodes()).toHaveLength(2)
      expect(await graph.nodes()).toContain("a")
      expect(await graph.nodes()).toContain("b")
      expect(await graph.serialize()).toMatchSnapshot()
    })

    it("Should remove nodes.", async () => {
      const graph = new Graph()
      await graph.addNode("a")
      await graph.addNode("b")
      await graph.removeNode("a")
      await graph.removeNode("b")
      expect(await graph.nodes()).toHaveLength(0)
    })

    it("Should add edges and query for adjacent nodes.", async () => {
      const graph = new Graph()
      await graph.addNode("a")
      await graph.addNode("b")
      await graph.addEdge("a", "b")
      expect(await graph.adjacent("a")).toHaveLength(1)
      expect(await graph.adjacent("a")).toMatchObject(["b"])
      expect(await graph.serialize()).toMatchSnapshot()
    })

    it("Should implicitly add nodes when edges are added.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")

      expect(await graph.adjacent("a")).toHaveLength(1)
      expect(await graph.adjacent("a")).toMatchObject(["b"])
      expect(await graph.nodes()).toHaveLength(2)
      expect(await graph.nodes()).toContain("a")
      expect(await graph.nodes()).toContain("b")
    })

    it("Should remove edges.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.removeEdge("a", "b")
      expect(await graph.adjacent("a")).toHaveLength(0)
    })

    it("Should not remove nodes when edges are removed.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.removeEdge("a", "b")
      expect(await graph.nodes()).toHaveLength(2)
      expect(await graph.nodes()).toContain("a")
      expect(await graph.nodes()).toContain("b")
    })

    it("Should remove outgoing edges when a node is removed.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.removeNode("a")
      expect(await graph.adjacent("a")).toHaveLength(0)
    })

    it("Should remove incoming edges when a node is removed.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.removeNode("b")
      expect(await graph.adjacent("a")).toHaveLength(0)
    })

    it("Should compute indegree.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      expect(await graph.indegree("a")).toBe(0)
      expect(await graph.indegree("b")).toBe(1)

      await graph.addEdge("c", "b")
      expect(await graph.indegree("b")).toBe(2)
    })

    it("Should compute outdegree.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      expect(await graph.outdegree("a")).toBe(1)
      expect(await graph.outdegree("b")).toBe(0)

      await graph.addEdge("a", "c")
      expect(await graph.outdegree("a")).toBe(2)
    })

  })

  describe("Algorithms", () => {

    // This example is from Cormen et al. "Introduction to Algorithms" page 550
    it("Should compute topological sort.", async () => {

      const graph = new Graph()

      // Shoes depend on socks.
      // Socks need to be put on before shoes.
      await graph.addEdge("socks", "shoes")

      await graph.addEdge("shirt", "belt")
      await graph.addEdge("shirt", "tie")
      await graph.addEdge("tie", "jacket")
      await graph.addEdge("belt", "jacket")
      await graph.addEdge("pants", "shoes")
      await graph.addEdge("underpants", "pants")
      await graph.addEdge("pants", "belt")

      const sorted = await graph.topologicalSort()

      expect(comesBefore(sorted, "pants", "shoes")).toBe(true)
      expect(comesBefore(sorted, "underpants", "pants")).toBe(true)
      expect(comesBefore(sorted, "underpants", "shoes")).toBe(true)
      expect(comesBefore(sorted, "shirt", "jacket")).toBe(true)
      expect(comesBefore(sorted, "shirt", "belt")).toBe(true)
      expect(comesBefore(sorted, "belt", "jacket")).toBe(true)

      expect(sorted).toHaveLength(8)

      expect(await graph.serialize()).toMatchSnapshot()
    })

    it("Should compute topological sort, excluding source nodes.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.addEdge("b", "c")
      const sorted = await graph.topologicalSort(["a"], false)
      expect(sorted).toHaveLength(2)
      expect(sorted[0]).toBe("b")
      expect(sorted[1]).toBe("c")
      expect(await graph.serialize()).toMatchSnapshot()
    })

    it("Should compute topological sort tricky case.", async () => {
      const graph = new Graph()     //      a
      //     / \
      await graph.addEdge("a", "b") //    b   |
      await graph.addEdge("a", "d") //    |   d
      await graph.addEdge("b", "c") //    c   |
      await graph.addEdge("d", "e") //     \ /
      await graph.addEdge("c", "e") //      e

      const sorted = await graph.topologicalSort(["a"], false)
      assert.equal(sorted.length, 4)
      assert(contains(sorted, "b"))
      assert(contains(sorted, "c"))
      assert(contains(sorted, "d"))
      assert.equal(sorted[sorted.length - 1], "e")

      assert(comesBefore(sorted, "b", "c"))
      assert(comesBefore(sorted, "b", "e"))
      assert(comesBefore(sorted, "c", "e"))
      assert(comesBefore(sorted, "d", "e"))

      expect(await graph.serialize()).toMatchSnapshot()
    })

    it("Should exclude source nodes with a cycle.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.addEdge("b", "c")
      await graph.addEdge("c", "a")

      const sorted = await graph.topologicalSort(["a"], false)
      assert.equal(sorted.length, 2)
      assert.equal(sorted[0], "b")
      assert.equal(sorted[1], "c")

      expect(await graph.serialize()).toMatchSnapshot()
    })

    it("Should exclude source nodes with multiple cycles.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.addEdge("b", "a")

      await graph.addEdge("b", "c")
      await graph.addEdge("c", "b")

      await graph.addEdge("a", "c")
      await graph.addEdge("c", "a")

      const sorted = await graph.topologicalSort(["a", "b"], false)
      assert(!contains(sorted, "b"))

      expect(await graph.serialize()).toMatchSnapshot()
    })

    it("Should compute lowest common ancestors.", async () => {
      const graph = new Graph()

      await graph.addEdge("a", "b")
      await graph.addEdge("b", "d")
      await graph.addEdge("c", "d")
      await graph.addEdge("b", "e")
      await graph.addEdge("c", "e")
      await graph.addEdge("d", "g")
      await graph.addEdge("e", "g")
      await graph.addNode("f")

      assert.deepStrictEqual(await graph.lowestCommonAncestors("a", "a"), ["a"])
      assert.deepStrictEqual(await graph.lowestCommonAncestors("a", "b"), ["b"])
      assert.deepStrictEqual(await graph.lowestCommonAncestors("a", "c"), ["d", "e"])
      assert.deepStrictEqual(await graph.lowestCommonAncestors("a", "f"), [])
    })
  })

  describe("Edge cases and error handling", () => {

    it("Should return empty array of adjacent nodes for unknown nodes.", async () => {
      const graph = new Graph()
      assert.equal((await graph.adjacent("a")).length, 0)
      assert.equal((await graph.nodes()).length, 0)
    })

    it("Should do nothing if removing an edge that does not exist.", async () => {
      const graph = new Graph()
      await graph.removeEdge("a", "b")
    })

    it("Should return indegree of 0 for unknown nodes.", async () => {
      const graph = new Graph()
      assert.equal(await graph.indegree("z"), 0)
    })

    it("Should return outdegree of 0 for unknown nodes.", async () => {
      const graph = new Graph()
      assert.equal(await graph.outdegree("z"), 0)
    })

  })

  describe("Serialization", () => {
    let serialized: Serialized

    function checkSerialized(graph: Serialized) {
      assert.equal(graph.nodes.length, 3)
      assert.equal(graph.links.length, 2)

      assert.equal(graph.nodes[0].id, "a")
      assert.equal(graph.nodes[1].id, "b")
      assert.equal(graph.nodes[2].id, "c")

      assert.equal(graph.links[0].source, "a")
      assert.equal(graph.links[0].target, "b")
      assert.equal(graph.links[1].source, "b")
      assert.equal(graph.links[1].target, "c")
    }

    it("Should serialize a await graph.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.addEdge("b", "c")
      serialized = await graph.serialize()
      checkSerialized(serialized)
    })

    it("Should deserialize a await graph.", async () => {
      const graph = new Graph()
      await graph.deserialize(serialized)
      checkSerialized(await graph.serialize())
    })

    it("Should chain deserialize a await graph.", async () => {
      const graph = new Graph()
      await graph.deserialize(serialized)
      checkSerialized(await graph.serialize())
    })
  })

  describe("Edge Weights", () => {

    it("Should set and get an edge weight.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b", 5)
      assert.equal(await graph.getEdgeWeight("a", "b"), 5)
    })

    it("Should set edge weight via setEdgeWeight.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.setEdgeWeight("a", "b", 5)
      assert.equal(await graph.getEdgeWeight("a", "b"), 5)
    })

    it("Should return weight of 1 if no weight set.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      assert.equal(await graph.getEdgeWeight("a", "b"), 1)
    })

  })

  describe("Dijkstra's Shortest Path Algorithm", () => {

    it("Should compute shortest path on a single edge.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      assert.deepEqual(await graph.shortestPath("a", "b"), withWeight(["a", "b"], 1))
    })

    it("Should compute shortest path on two edges.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.addEdge("b", "c")
      assert.deepEqual(await graph.shortestPath("a", "c"), withWeight(["a", "b", "c"], 2))
    })

    it("Should compute shortest path on example from Cormen text (p. 659).", async () => {
      const graph = new Graph()
      await graph.addEdge("s", "t", 10)
      await graph.addEdge("s", "y", 5)
      await graph.addEdge("t", "y", 2)
      await graph.addEdge("y", "t", 3)
      await graph.addEdge("t", "x", 1)
      await graph.addEdge("y", "x", 9)
      await graph.addEdge("y", "z", 2)
      await graph.addEdge("x", "z", 4)
      await graph.addEdge("z", "x", 6)

      assert.deepEqual(await graph.shortestPath("s", "z"), withWeight(["s", "y", "z"], 5 + 2))
      assert.deepEqual(await graph.shortestPath("s", "x"), withWeight(["s", "y", "t", "x"], 5 + 3 + 1))
    })

    it("Should throw error if source node not in await graph.", async () => {
      const graph = new Graph()
      await graph.addEdge("b", "c")
      expect(graph.shortestPath("a", "c")).rejects.toThrowError(/Source node/)
    })

    it("Should throw error if dest node not in await graph.", async () => {
      const graph = new Graph()
      await graph.addEdge("b", "c")
      expect(graph.shortestPath("b", "g")).rejects.toThrowError(/Destination node/)
    })

    it("Should throw error if no path exists.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.addEdge("d", "e")
      expect(graph.shortestPath("a", "e")).rejects.toThrowError(/No path/)
    })

    it("Should be robust to disconnected subgraphs.", async () => {
      const graph = new Graph()
      await graph.addEdge("a", "b")
      await graph.addEdge("b", "c")
      await graph.addEdge("d", "e")
      const shortedPath = await graph.shortestPath("a", "c")
      assert.deepEqual(shortedPath, withWeight(["a", "b", "c"], 2))
    })
  })
})
