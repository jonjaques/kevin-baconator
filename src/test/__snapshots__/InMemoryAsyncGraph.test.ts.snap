// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Algorithms Should compute topological sort tricky case. 1`] = `
Object {
  "links": Array [
    Object {
      "source": "a",
      "target": "b",
      "weight": 1,
    },
    Object {
      "source": "a",
      "target": "d",
      "weight": 1,
    },
    Object {
      "source": "b",
      "target": "c",
      "weight": 1,
    },
    Object {
      "source": "d",
      "target": "e",
      "weight": 1,
    },
    Object {
      "source": "c",
      "target": "e",
      "weight": 1,
    },
  ],
  "nodes": Array [
    Object {
      "id": "a",
    },
    Object {
      "id": "b",
    },
    Object {
      "id": "d",
    },
    Object {
      "id": "c",
    },
    Object {
      "id": "e",
    },
  ],
}
`;

exports[`Graph Algorithms Should compute topological sort, excluding source nodes. 1`] = `
Object {
  "links": Array [
    Object {
      "source": "a",
      "target": "b",
      "weight": 1,
    },
    Object {
      "source": "b",
      "target": "c",
      "weight": 1,
    },
  ],
  "nodes": Array [
    Object {
      "id": "a",
    },
    Object {
      "id": "b",
    },
    Object {
      "id": "c",
    },
  ],
}
`;

exports[`Graph Algorithms Should compute topological sort. 1`] = `
Object {
  "links": Array [
    Object {
      "source": "socks",
      "target": "shoes",
      "weight": 1,
    },
    Object {
      "source": "shirt",
      "target": "belt",
      "weight": 1,
    },
    Object {
      "source": "shirt",
      "target": "tie",
      "weight": 1,
    },
    Object {
      "source": "belt",
      "target": "jacket",
      "weight": 1,
    },
    Object {
      "source": "tie",
      "target": "jacket",
      "weight": 1,
    },
    Object {
      "source": "pants",
      "target": "shoes",
      "weight": 1,
    },
    Object {
      "source": "pants",
      "target": "belt",
      "weight": 1,
    },
    Object {
      "source": "underpants",
      "target": "pants",
      "weight": 1,
    },
  ],
  "nodes": Array [
    Object {
      "id": "socks",
    },
    Object {
      "id": "shoes",
    },
    Object {
      "id": "shirt",
    },
    Object {
      "id": "belt",
    },
    Object {
      "id": "tie",
    },
    Object {
      "id": "jacket",
    },
    Object {
      "id": "pants",
    },
    Object {
      "id": "underpants",
    },
  ],
}
`;

exports[`Graph Algorithms Should exclude source nodes with a cycle. 1`] = `
Object {
  "links": Array [
    Object {
      "source": "a",
      "target": "b",
      "weight": 1,
    },
    Object {
      "source": "b",
      "target": "c",
      "weight": 1,
    },
    Object {
      "source": "c",
      "target": "a",
      "weight": 1,
    },
  ],
  "nodes": Array [
    Object {
      "id": "a",
    },
    Object {
      "id": "b",
    },
    Object {
      "id": "c",
    },
  ],
}
`;

exports[`Graph Algorithms Should exclude source nodes with multiple cycles. 1`] = `
Object {
  "links": Array [
    Object {
      "source": "a",
      "target": "b",
      "weight": 1,
    },
    Object {
      "source": "a",
      "target": "c",
      "weight": 1,
    },
    Object {
      "source": "b",
      "target": "a",
      "weight": 1,
    },
    Object {
      "source": "b",
      "target": "c",
      "weight": 1,
    },
    Object {
      "source": "c",
      "target": "b",
      "weight": 1,
    },
    Object {
      "source": "c",
      "target": "a",
      "weight": 1,
    },
  ],
  "nodes": Array [
    Object {
      "id": "a",
    },
    Object {
      "id": "b",
    },
    Object {
      "id": "c",
    },
  ],
}
`;

exports[`Graph Data structure Should add edges and query for adjacent nodes. 1`] = `
Object {
  "links": Array [
    Object {
      "source": "a",
      "target": "b",
      "weight": 1,
    },
  ],
  "nodes": Array [
    Object {
      "id": "a",
    },
    Object {
      "id": "b",
    },
  ],
}
`;

exports[`Graph Data structure Should add nodes and list them. 1`] = `
Object {
  "links": Array [],
  "nodes": Array [
    Object {
      "id": "a",
    },
    Object {
      "id": "b",
    },
  ],
}
`;
